hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -978,3 +997,144 @@ def array_for(index_shape, size):
+    if len(args) <= 1:
+        # Zero args: ndarray, _SupportsArray
+        # One arg: ndarray[type], _SupportsArray[type]
+        shape = Any
+        dtype = _unpack_dtype(args[0]) if args else Any
+    else:
+        # Two args: ndarray[shape, type], NDArray[*]
+        assert len(args) == 2
+        shape = args[0]
+        assert shape is Any
+        dtype = _unpack_dtype(args[1])
+    return (
+        scalar_dtypes() if dtype is Any else np.dtype(dtype),
+        array_shapes(max_dims=2) if shape is Any else shape,
+    )
+
+
+def _from_type(thing: Type[Ex]) -> Optional[st.SearchStrategy[Ex]]:
+    """Called by st.from_type to try to infer a strategy for thing using numpy.
+
+    If we can infer a numpy-specific strategy for thing, we return that; otherwise,
+    we return None.
+    """
+
+    base_strats = st.one_of(
+        [
+            st.booleans(),
+            st.integers(),
+            st.floats(),
+            st.complex_numbers(),
+            st.text(),
+            st.binary(),
+        ]
+    )
+    # np.array(arr_like) (1.24.3) fails if mixing strings and non-ascii
+    # bytestrings (ex: ['', b'\x80'])
+    base_strats_ascii = st.one_of(
+        [
+            st.booleans(),
+            st.integers(),
+            st.floats(),
+            st.complex_numbers(),
+            st.text(),
+            st.binary().filter(bytes.isascii),
+        ]
+    )
+
+    if thing == np.dtype:
+        return array_dtypes()
+
+    if thing == ArrayLike:
+        # We override the default type resolution to ensure the "coercible to
+        # array" contract is honoured. See
+        # https://github.com/HypothesisWorks/hypothesis/pull/3670#issuecomment-1578140422.
+        # The actual type is (as of np 1.24), with
+        # scalars:=[bool, int, float, complex, str, bytes]:
+        # Union[
+        #     _SupportsArray,
+        #     _NestedSequence[_SupportsArray],
+        #     *scalars,
+        #     _NestedSequence[Union[*scalars]]
+        # ]
+        return st.one_of(
+            # *scalars
+            base_strats,
+            # _NestedSequence[Union[*scalars]], but excluding non-ascii binary
+            st.lists(base_strats_ascii),
+            # _SupportsArray, use plain ndarrays
+            st.from_type(np.ndarray),
+            # _NestedSequence[_SupportsArray], but guaranteeing equal size
+            st.integers(min_value=0, max_value=4).flatmap(
+                lambda s: st.one_of(
+                    st.recursive(
+                        st.lists(base_strats_ascii, min_size=s, max_size=s),
+                        extend=st.tuples,
+                    ),
+                    st.recursive(
+                        arrays(
+                            scalar_dtypes(),
+                            array_shapes(min_dims=s, max_dims=s, min_side=s, max_side=s),
+                        ),
+                        extend=st.tuples,
+                    ),
+                ),
+            ),
```

Comment:
This is my attempt at making an arbitrary nested structure of fixed dimension. Is there a (simple) better way?